File: datastruct/tree/trie/src/scala/PrefixTree.scala
import scala.util.Random    // for verification
import scala.language.postfixOps

object PrefixTree {
  case class Tree[+K, V] (v: Option[V], cs: List[(List[K], Tree[K, V])]) {
    val value = v
    val children = cs
  }

  def empty[K, V](): Tree[K, V] = Tree(None, List())

  def leaf[K, V](v: V): Tree[K, V] = Tree(Some(v), List())

  def insert[K, V] (t: Tree[K, V], key: List[K], v: V): Tree[K, V] = {
    def ins[K, V] (ts: List[(List[K], Tree[K, V])],
                   key: List[K], v: V): List[(List[K], Tree[K, V])] =
      ts match {
        case List() => List((key, leaf(v)))
        case p :: ps => {
          val (key1, t1) = p
          if (key1 == key) {
            (key, Tree(Some(v), t1.children)) :: ps  // overwrite
          } else if (hasPrefix(key1, key)) {
            branch(key, v, key1, t1) :: ps
          } else {
            p :: ins(ps, key, v)
          }
        }
      }
    Tree(t.value, ins(t.children, key, v))
  }

  def hasPrefix[K] (key1: List[K], key2: List[K]) =
    ! key1.isEmpty && ! key2.isEmpty && key1.head == key2.head

  def branch[K, V] (key1: List[K], v: V,
                    key2: List[K], t2: Tree[K, V]): (List[K], Tree[K, V]) = {
    val key = lcp(key1, key2)
    val m = key.length
    val newKey1 = key1.drop(m)
    val newKey2 = key2.drop(m)
    if (key1 == key) { // insert "an" into "another"
      (key, Tree(Some(v), List((newKey2, t2))))
    } else if (key2 == key) { // insert "another" into "an"
      (key, insert(t2, newKey1, v))
    } else {
      (key, Tree(None, List((newKey1, leaf(x)), (newKey2, t2))))
    }
  }

  // the longest common prefix
  def lcp[K] (xs: List[K], ys: List[K]): List[K] =
    if (!xs.isEmpty && !ys.isEmpty && xs.head == ys.head) {
      xs.head :: lcp(xs.tail, ys.tail)
    } else {
      List()
    }

  // lookup
  def lookup[K, V] (t: Tree[K, V], key: List[K]): Option[V] = {
    def extract[K] (xs: List[K], ys: List[K]) = ys.drop(lcp(xs, ys).length)
    def find[K, V] (assoc: List[(List[K], Tree[K, V])], key: List[K]): Option[V] =
      assoc match {
        case List() => None
        case p :: ps => {
          val (key1, t1) = p
          if (key1 == key) {
            t1.value
          } else if (ks.startsWith(key1)) {
            lookup(t1, extract(key, key1))
          }
        }
      }
    find(t.children, key)
  }
}

File: datastruct/tree/trie/src/PrefixTree.hs
{-
    PrefixTree.hs, Alphabetic Prefix Tree.
    Although it's called 'alphabetic', we abstract the key as sequence
    Copyright (C) 2010, Liu Xinyu (liuxinyu95@gmail.com)

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
-}

module PrefixTree where

import Data.List (isPrefixOf, sort, sortBy)
import Data.Function (on)

-- definition
data PrefixTree k v = PrefixTree { value :: Maybe v
                                 , children :: [([k], PrefixTree k v)]}

empty = PrefixTree Nothing []

leaf x = PrefixTree (Just x) []

-- insertion
insert :: Eq k => PrefixTree k v -> [k] -> v -> PrefixTree k v
insert t ks x = PrefixTree (value t) (ins (children t) ks x) where
    ins []     ks x = [(ks, PrefixTree (Just x) [])]
    ins (p@(ks', t') : ps) ks x
        | ks' == ks
            = (ks, PrefixTree (Just x) (children t')) : ps  -- overwrite
        | match ks' ks
            = (branch ks x ks' t') : ps
        | otherwise
            = p : (ins ps ks x)

match x y = x /= [] && y /= [] && head x == head y

branch :: Eq k => [k] -> v -> [k] -> PrefixTree k v -> ([k], PrefixTree k v)
branch ks1 x ks2 t2
    | ks1 == ks
        -- ex: insert "an" into "another"
        = (ks, PrefixTree (Just x) [(ks2', t2)])
    | ks2 == ks
        -- ex: insert "another" into "an"
        = (ks, insert t2 ks1' x)
    | otherwise = (ks, PrefixTree Nothing [(ks1', leaf x), (ks2', t2)])
   where
      ks = lcp ks1 ks2
      m = length ks
      ks1' = drop m ks1
      ks2' = drop m ks2

-- longest common prefix
lcp :: Eq k => [k] -> [k] -> [k]
lcp [] _ = []
lcp _ [] = []
lcp (x:xs) (y:ys) = if x==y then x : (lcp xs ys) else []

-- lookup
find :: Eq k => PrefixTree k v -> [k] -> Maybe v
find t ks = find' (children t) ks where
    find' [] _ = Nothing
    find' (p@(ks', t') : ps) k
          | ks' == ks = value t'
          | ks' `isPrefixOf` ks = find t' (diff ks' ks)
          | otherwise = find' ps ks
    diff ks1 ks2 = drop (length (lcp ks1 ks2)) ks2

fromList :: Eq k => [([k], v)] -> PrefixTree k v
fromList = foldl ins empty where
    ins t (k, v) = insert t k v

fromString :: (Enum v, Num v) => String -> PrefixTree Char v
fromString = fromList . (flip zip [1..]) . words

-- Pre-order traverse to populate keys in lexicographical order.
keys :: Ord k => PrefixTree k v -> [[k]]
keys = keys' [] where
  keys' prefix t = case (value t) of
    Nothing -> kss
    (Just _) -> prefix : kss
    where
      kss = concatMap (\(ks, t') -> keys' (prefix ++ ks) t') ts
      ts = sortBy (compare `on` fst) (children t)

--
example = insert (fromString "a place where animals are for public to see") "zoo" 0

-- test data
assocs = [[("a", 1), ("an", 2), ("another", 7), ("boy", 3), ("bool", 4), ("zoo", 3)],
          [("zoo", 3), ("bool", 4), ("boy", 3), ("another", 7), ("an", 2), ("a", 1)]]

verify = all (\as ->
                 let t = fromList as in
                   all (\(k, v) -> maybe False (==v) (find t k)) as) assocs

verifyKeys = all (\as ->
                   keys (fromList as) == (sort $ fst $ unzip as)) assocs
